kv/kvserver: TestSnapshotAfterTruncation failed

kv/kvserver.TestSnapshotAfterTruncation [failed](https://mesolite.cluster.engflow.com/invocations/default/941241aa-60a9-46a1-a9a7-bad87a719260?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [0ea9ae61bfe16ba95414c287753c67334e92e578](https://github.com/cockroachdb/cockroach/commits/0ea9ae61bfe16ba95414c287753c67334e92e578):


```
=== RUN   TestSnapshotAfterTruncation
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestSnapshotAfterTruncation3104363652
    test_log_scope.go:76: use -show-logs to present logs inline
    client_raft_test.go:828: -- test log scope end --
test logs left over in: outputs.zip/logTestSnapshotAfterTruncation3104363652
--- FAIL: TestSnapshotAfterTruncation (79.99s)
=== RUN   TestSnapshotAfterTruncation/differentTerm
    client_raft_test.go:760: condition failed to evaluate within 45s: from client_raft_test.go:780: no leader
    --- FAIL: TestSnapshotAfterTruncation/differentTerm (63.79s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSnapshotAfterTruncation.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvserver.TestSnapshotAfterTruncation [failed](https://tanzanite.cluster.engflow.com/invocations/default/19c59fd8-74b2-488c-98d1-c5bf93a4d383?testReportRun=15&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [b3fec61ca90095c664f6432af864f18e9946f8bb](https://github.com/cockroachdb/cockroach/commits/b3fec61ca90095c664f6432af864f18e9946f8bb):


```
=== RUN   TestSnapshotAfterTruncation
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestSnapshotAfterTruncation3965460839
    test_log_scope.go:76: use -show-logs to present logs inline
    client_raft_test.go:828: -- test log scope end --
test logs left over in: outputs.zip/logTestSnapshotAfterTruncation3965460839
--- FAIL: TestSnapshotAfterTruncation (70.48s)
=== RUN   TestSnapshotAfterTruncation/differentTerm
    client_raft_test.go:760: condition failed to evaluate within 45s: from client_raft_test.go:780: no leader
    --- FAIL: TestSnapshotAfterTruncation/differentTerm (61.56s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=15</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSnapshotAfterTruncation.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvserver.TestSnapshotAfterTruncation [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18022653?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18022653?buildTab=artifacts#/) on master @ [6a61ee06779f1e6e0fe68be4286cf4cac3e8aa4e](https://github.com/cockroachdb/cockroach/commits/6a61ee06779f1e6e0fe68be4286cf4cac3e8aa4e):


```
=== RUN   TestSnapshotAfterTruncation
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/37ddc8e37e5448832e32de57f39c38ec/logTestSnapshotAfterTruncation4140550560
    test_log_scope.go:76: use -show-logs to present logs inline
    client_raft_test.go:828: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/37ddc8e37e5448832e32de57f39c38ec/logTestSnapshotAfterTruncation4140550560
--- FAIL: TestSnapshotAfterTruncation (76.11s)
=== RUN   TestSnapshotAfterTruncation/differentTerm
    client_raft_test.go:760: condition failed to evaluate within 45s: from client_raft_test.go:780: no leader
    --- FAIL: TestSnapshotAfterTruncation/differentTerm (67.78s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSnapshotAfterTruncation.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -775,6 +775,17 @@ func TestSnapshotAfterTruncation(t *testing.T) {
 						} else if status.Term != term {
 							return errors.Errorf("terms do not agree: %d vs %d", status.Term, term)
 						}
+						if !hasLeader {
+							// If we haven't been able to establish a  leader yet, send a get
+							// request to force the issue.
+							getReq := getArgs(key)
+							_, err := kv.SendWrapped(ctx, tc.GetFirstStoreFromServer(t, i).TestSender(), getReq)
+							// It should be fine if we get a NLHE for some reason.
+							nlhe := &kvpb.NotLeaseHolderError{}
+							if !errors.As(err.GetDetail(), &nlhe) {
+								return err.GetDetail()
+							}
+						}
 					}
 					if !hasLeader {
 						return errors.New("no leader")

